### DISCUSSION TITLE
How to use Assert\When in array data validation

### DISCUSSION THREAD

**Discussion by QuietNoise** [symfony/symfony#49112] [Q&A] [answered]
I'm validating contents of an associative array and I can't figure out how to access values of other fields in **Assert\When** expression. 
See the simplified example below. I need validate **blob** only when **hasBlob** is set to true. 
I'm trying to figure out how to access value **hasBlob** from within the **When** expression. 
Apparently **this.hasBlob**  won't work because I don't validate object and **this** is null. And I can't use ` 'expression' => 'value` because **value** has value of the **blob** and there is no access to other fields.
I would appreciate if someone can point me in the right direction. 

_Note: Keep in mind that I'm using Assert\Collection data structure for contraints here rather than hardcoding constraints class because fields and their constraints will be determined during runtime. For simplicity I just show one of the most basic scenarios._

```
<?php
namespace Example;

use Symfony\Component\Validator\Constraints as Assert;
use Symfony\Component\Validator\ConstraintViolation;
use Symfony\Component\Validator\Validation;


$data = [
            'title' => 'Oz',
            'hasBlob' => true,
            'blob' => ''
        ];

$constraints = new Assert\Collection([
    'fields' => [
        'title' => [
            new Assert\Type('string'),
            new Assert\Length(['min' => 0, 'max' => 100]),
        ],
        'hasBlob' => [
            new Assert\Type('boolean')
        ],
        'blob' => [
            new Assert\When([                
                'expression' => 'this.hasBlob === true',
                'constraints' => [
                    new Assert\NotBlank(),
                    new Assert\Type('string'),
                    new Assert\Length(['min' => 1, 'max' => 1000]),
                ],
            ])
        ]
    ]
]);

$validator = Validation::createValidator();

$violations = $validator->validate($data, $constraints);
var_dump($violations);
```

**Comment by MatTheCat** [comment, marked as answer]
Oh this is weird, `this` comes from the context’s object, so if there is no object (like in your case) you get `null`. I’ll dig into this but in the meantime you can use a `Callback` constraint to validate the field based on others:

```diff
'blob' => [
    new Assert\Callback(static function (mixed $blob, ExecutionContextInterface $context): void
    {
        $root = $context->getRoot();
        if (!$root['hasBlob']) {
            return;
        }

        $context->getValidator()->inContext($context)->validate($blob, [
            new Assert\NotBlank(),
            new Assert\Type('string'),
            new Assert\Length(['min' => 1, 'max' => 1000]),
        ]);
    })
]
```

Also note that `hasBlob` can be computed from `blob` so maybe you don’t even need all of this.

**Reply by MatTheCat** [reply]
In fact `When` has a `$values` parameter you can use to inject variables in the expression:

```php
'blob' => [
    new Assert\When([
         'expression' => 'hasBlob === true',
         'constraints' => [
            new Assert\NotBlank(),
            new Assert\Type('string'),
            new Assert\Length(['min' => 1, 'max' => 1000]),
        ],
        'values' => ['hasBlob' => $data['hasBlob']]
    ])
]
```

Still,

> note that `hasBlob` can be computed from `blob` so maybe you don’t even need all of this.

**Reply by QuietNoise** [reply]
Thank you,
The reason I want use `When` is because the code is much more readable. I played with injection just now and I believe this is the most elegant way to do it for now. 
While I can compute `hasBlob` from `blob ` itselt in the example above this is just a simplified example. 
In real life I need compute more complex conditions and based on that decide format of constraints for depended fields.
Thank you for your help.
